nx/lib/nx.ex

```diff
@@ -14124,13 +14124,17 @@ defmodule Nx do
     else
       tensor = devectorize(tensor, keep_names: false)
       indices = devectorize(indices, keep_names: false)
+      gather_indices = reshape(indices, indices |> shape |> Tuple.append(1))
 
-      impl!(tensor).take(
-        %{tensor | shape: inner_shape, names: inner_names},
-        tensor,
-        indices,
-        axis
-      )
+      {indices_axes, tensor_axes} = Enum.split(axes(inner_shape), rank(indices))
+      {leading, trailing} = Enum.split(tensor_axes, axis)
+
+      transpose_axes = leading ++ indices_axes ++ trailing
+
+      tensor
+      |> gather(gather_indices, axes: [axis])
+      |> transpose(axes: transpose_axes)
```

Comment:
Why is this transpose needed?

Reply:
I took a look at this. `take` will retain the shape according to the given axis, this is also true in Python:

```
>>> np.take(n, [0, 1, 0, 1], axis=1)
array([[1, 2, 1, 2],
       [3, 4, 3, 4]])
>>> np.take(n, [0, 1, 0, 1], axis=0)
array([[1, 2],
       [3, 4],
       [1, 2],
       [3, 4]])
```

On the other hand, gather will always preserve the input shape. This means the underlying `offset_dims` in the gather operation is different. So it may be that we cannot really replace take by gather (unless we want to use lose performance when axis != 0, which may be fine). Another option is to extend gather with a new option that controls the output shape but that's debatable.

Reply:
I will merge this and make it optional in a follow up commit.